File: Valid Mountain Array.py
# https://leetcode.com/problems/valid-mountain-array/

class Solution:
    def validMountainArray(self, arr: List[int]) -> bool:
        
        # If arr's length is less than 3, return false
        if len(arr) < 3:
            return False
        
        # Create a pointer and get size of list
        peak = 0
        arrSize = len(arr)
        
        # Traverse array from left-to-right
        # I.e., Walk up the mountain
        while peak + 1 < arrSize and arr[peak] < arr[peak + 1]:
            peak += 1
        
        # If peak element is either the first or last element, return false
        # E.g., 0 1 2 3 4 or 4 3 2 1
        if peak == 0 or peak == len(arr) - 1:
            return False
        
        # Traverse array from right-to-left
        # I.e., Walk down the mountain
        while peak + 1 < arrSize and arr[peak] > arr[peak + 1]:
            peak += 1
        
        # If pointer reached the end of the list (climbed down the mountain), return true
        return peak == len(arr) - 1

File: Valid Mountain Array.java
// https://leetcode.com/problems/valid-mountain-array/

class Solution {
    public boolean validMountainArray(int[] arr) {
        
        // If array's size is less than 3, return false
        if (arr.length < 3) {return false;}
        
        // Create a pointer and get array's size
        int peak = 0;
        int arrSize = arr.length;
        
        // Traverse the array from left-to-right
        // I.e., Climb up the mountain
        while (peak + 1 < arrSize && arr[peak] < arr[peak + 1]) {
            peak++;
        }
        
        // If pointer is either at the beginning or the end of the array, return false
        // I.e., Peak element can't be at either endpoints, or else it's not a mountain array
        if (peak == 0 || peak == arrSize - 1) {
            return false;
        }
        
        // Continue to traverse the array
        // I.e., Climb down the mountain
        while (peak + 1 < arrSize && arr[peak] > arr[peak + 1]) {
            peak++;
        }
        
        // If mountain climber (pointer) reached the end of the mountain, return true
        return peak == arrSize - 1;
    }
}
